django_saml2_auth/views.py

```diff
@@ -97,9 +98,16 @@ def acs(request: HttpRequest):
     use_jwt = dictor(saml2_auth_settings, "USE_JWT", False)
     if use_jwt and target_user.is_active:
         # Create a new JWT token for IdP-initiated login (acs)
-        jwt_token = create_jwt_token(target_user.email)
-        # Use JWT auth to send token to frontend
-        query = f"?token={jwt_token}"
+
+        # Check if there is a custom trigger for creating the JWT and URL query
+        custom_jwt_query_trigger = dictor(saml2_auth_settings, "TRIGGER.CUSTOM_TOKEN_QUERY")
+        if custom_jwt_query_trigger:
+            query = run_hook(custom_jwt_query_trigger, target_user)
```

Comment:
How about [decoding](https://github.com/grafana/django-saml2-auth/blob/9213beeabc17e0990c6c08d15e8efa450bc5fff1/django_saml2_auth/views.py#L137)?

Reply:
Actually I wasn't using the sp_initiated_login view but only "signin" and "acs"; what should be the use case of the sp initiated login? I mean, where does the token in the GET method come from?

Reply:
It usually comes from a redirect in your webapp.

Reply:
So, at the moment the flow I am using is the following:
- frontend webapp calls the backend django rest api at django-saml2-auth "signin" view
- backend creates an AuthNRequest and redirect to IdP
- user login at the IdP and is redirected to backend
- backend confirm AuthNResponse, creates a JWT token and redirects the user to the frontend webapp with the token to be used in the following api backend calls

In the sp_initiated_login the case could be that the frontend webapp is calling the login with a token to be checked or refreshed by the IdP?

However, probably to make it more customisable we could add another trigger function for checking and refreshing the token?
In the view it gets the user_id from the token, prepare saml, recreate a token and pass it to the RelayState:
```python
user_id = decode_jwt_token(request.GET.get("token"))
saml_client = get_saml_client(get_assertion_url(request), acs, user_id)
jwt_token = create_jwt_token(user_id)
```

How do you think would be a good way to do that hook?

Reply:
@sgabb 

In the SP-initiated login, this is the flow:
- Frontend sends a login request to the backend (usually just an email).
- Backend verifies the requests and returns a redirect URL with a signed JWT token to the frontend.
- Frontend redirects the user to the URL along with the token. The URL points to the sp_initiated login endpoint.
- The endpoint refreshes the token and then the user is redirected to the IdP login page with the new token in their AuthNRequest.
- After login, the flow is the same as the IdP-initiated login that goes through the `/sso/acs` endpoint.

I suggest that if we want to decode the token somewhere, just like encoding, we should consider having another hook that deals with decoding custom tokens. The `create_jwt_token` could make use of the `TRIGGER.CUSTOM_TOKEN_QUERY` hook and probably be renamed to something more meaningful, like `CUSTOM_CREATE_JWT_TOKEN`, and then the `decode_jwt_token` hook can use another hook with the reverse logic, called `CUSTOM_DECODE_JWT_TOKEN`, which of course, should be documented. The names are just suggestions, so feel free to use any name that makes more sense.

Reply:
@mostafa 
I am working on it.
Quick question: how should I run the test in development?
I tried just running `pytest` but I get _"No module named 'django_saml2_auth.tests.settings'"_ :sweat: